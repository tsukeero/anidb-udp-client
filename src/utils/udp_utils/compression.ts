import zlib from 'node:zlib'
// import * as streamConsumers from 'node:stream/consumers'
// import { PassThrough } from 'node:stream'

// TODO: figure out why EOF, and way to fix it
export const decompress = (buffer: Buffer): Promise<Buffer> => new Promise((resolve, reject) => {
  const buffer_builder: Buffer[] = []
  const decompress_stream = zlib.createUnzip()
    .on('data', (chunk: Buffer) => {
      buffer_builder.push(chunk)
    }).on('close', () => {
      resolve(Buffer.concat(buffer_builder))
    }).on('error', (err: any) => {
      if (err.errno !== -5) // EOF: expected
        reject(err)
    })

  decompress_stream.write(buffer.subarray(2))
  decompress_stream.end()
})

// export const decompress = async (buffer: Buffer): Promise<Buffer> => {
//   const decompress_stream = zlib.createUnzip()
// //   const stream = new PassThrough()
// //   decompress_stream.on('error', function (err: any) {
// //     if (err.code === 'Z_BUF_ERROR') {
// //       stream.end()
// //       return
// //     }
// //     stream.emit('error', err)
// //   })
// //   decompress_stream.pipe(stream)

//   decompress_stream.write(buffer.subarray(2))
//   decompress_stream.end()
// //   decompress_stream.flush()
//   const data = await streamConsumers.buffer(decompress_stream)
//   return data
// }
export default decompress