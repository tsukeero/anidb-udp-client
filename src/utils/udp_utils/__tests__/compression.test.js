import { decompress } from '../compression'
const SAMPLE_COMPRESSED_DATA = Buffer.from('AAB4nEWUy27lKBCG9/0U7LKpSAZ8wct0J4seTXePND29JzbHRoPBDZxEls7Dz19Ootl8LqAo/rrIr/3xTxRKN+Lh+9dvT5862chbZ1p517aNupO96tubvNMwm+b2JdgY7Xz78ufD9+8Pjx9r8fPX3Ycpb1KOsmtVC//BNONNK2Okvm3Wx+qijZMT1S6FrnF2dqbg6mqDmFZ3odUvqyjTmlKg2QUff19drLSsqVSaEt+vwsfZT7amXGg+ot38RNXmxVVhr7N3iE91dZsrlLJffETw15T/peJq9XEhF3AYa6HqN0d7sLiwZ1f4pYufqk98JSS8AZOebK7r2+PbnrLNh7jYWG05qAS7l+onxD41iylcnwv9YXcb6cWncIopIb2K19VF4auYEoSJmhD/xaEE2f2++urmtzVD7CkcCx7eQ2L/CNFwwWvVZZfqsSOz3eVLyhsXkxaLNF58uUJ1RIBAD8Vb2m22Ibgz+TCTjQtq+KGcZcwH5fQWYrXZbXTJqN5cVr/TbH04RPAX95HaaaOuAuVA6bJdOADioKQiXQTWSKekjGQoOxvuz3f/r+NmA3LLCb1P0UOLndEfXIams3C4tdkzyVm8uFxwSdgXCLHPwfE42AMnsr8fxXNOdp7gTPkaIwdZ7EI+hOhKobLb6VR1rdzYmTUibyR0caeK8+ZzOsTmXC1i8TmgBPt+COTPI7I6ZHn/ajNnhyRDuma0YkZqGGI7VYjeUCF6+4hXX1e8JraECmUq/jmwgMNHcaDwtLmQ5mw3iypE+FfEL+dg3EMrorKEj+2fP8TnJ/Htx6+nRwH76a+vf/94fKLqpjVi8IOwZXdT/RhvOrOpOfFcvOt/l/XWPzhfA3qA0Xof77cJxcRzP+bruXfTDUnZKZLtIEk1rQb6kVSnO1J90zJOq2cMDJz2smFIhmLgWq/YWRlA82nLVscHHK8f4Dd0OBgGHAymYyDoMPLeKBlwNk3DQCijT8DZtGy1iGI4nul4r+Nlrxh8ym8Yw6dmYMBl1JJ0A0G6aRvSqlGkh7EjPQ4dtUppYOyp1WxpDeDnBxjGaKgd1Ejd0BsAIvF71ACeBDrAtNSNQC+Rai87CQyaevxWGWy1PWOkoUHmgzHdTXZaoz7NON6k7lAUqQcoZpqTI1phUAlmC+KvSm3TInCLIPKkOnnutCcN7/esWfKtYexRWubAHE97ZHvoTvY96U7iWYnOAT3jXBoGNEjFg6EkQzHYDx0GOgY7K3ZW7IxWsXYeI/OeQXNSnlSf/gPedg==','base64')

describe('compression', () => {
  test('decompress',async () => {
    const decompressed = await decompress(SAMPLE_COMPRESSED_DATA)
    expect(decompressed.toString('utf-8')).toMatchSnapshot()
  })
})