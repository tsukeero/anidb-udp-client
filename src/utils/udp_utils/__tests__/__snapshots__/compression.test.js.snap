// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compression decompress 1`] = `
"w6yUn 230 ANIME
5101|5841'4402'16264|1'32'100|Clannad|CLANNAD|Clannad TV'Clannad 1|1191542400|7809|328813|maintenance tags,undead,lethal chef,high school,delinquent,ghost,content indicators,dynamic,target audience,themes,original work,setting,elements,time,place,present,fictional location,Earth,contemporary fantasy,slapstick,school clubs,Japan,violence,slow when it comes to love,unrequited love,love polygon,plot continuity,stereotypes,performance,game,visual novel,Asia,parallel world,angst,fantasy,comedy,romance,harem,friendship,daily life,school life,the arts,tragedy,coming of age,unsorted,real-world location,male protagonist,adoring fan club,remastered version available,delayed 16-9 broadcast,running gag,illness,spacing out,predominantly female cast,boy meets girl,happy ending,heart-warming,colour coded,intact family,family without mother,sibling yin yang,melodrama,manly tears,slow-paced,girly tears,TO BE MOVED TO EPISODE,technical aspects,origin,cast,tropes,ending,family life,speculative fiction,Japanese production|30,1152,1471,2043,2069,2535,2604,2605,2606,2607,2609,2610,2611,2612,2613,2624,2628,2630,2648,2653,2669,2672,2750,2778,2785,2786,2790,2791,2793,2800,2804,2834,2838,2844,2849,2853,2858,2859,2862,2864,2869,2883,2887,2888,2931,3028,3040,3202,3795,3975,4223,4296,4323,4333,4440,4480,4498,4729,5768,5790,5843,5862,5865,5884,5984,6148,6151,6173,6230,6233,6246,6249,7050,7885|1533630099|13572,13787,13788,13789,14893,14894,14915,40473,40501,40502,40503,40543,40581,40698,41894,79686,79687,79696,79697,79757,79766,3517,1113,1116,1117,1118,1119,1120,1121,1122,1123,1124,1125,1127,1128,1129,1134,13782,14889,14890,14891,14892
"
`;
